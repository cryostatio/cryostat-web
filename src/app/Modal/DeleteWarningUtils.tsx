/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
export enum DeleteWarningType {
  DeleteActiveRecordings = 'DeleteActiveRecordings',
  DeleteArchivedRecordings = 'DeleteArchivedRecordings',
  DeleteAutomatedRules = 'DeleteAutomatedRules',
  DeleteEventTemplates = 'DeleteEventTemplates',
  DeleteProbeTemplates = 'DeleteProbeTemplates',
  DeleteActiveProbes = 'DeleteActiveProbes',
  DeleteJMXCredentials = 'DeleteJMXCredentials',
  DeleteCustomTargets = 'DeleteCustomTargets',
}

export interface DeleteWarning {
  id: DeleteWarningType;
  title: string;
  label: string;
  description: string;
  ariaLabel: string;
}

export const DeleteActiveRecordings: DeleteWarning = {
  id: DeleteWarningType.DeleteActiveRecordings,
  title: 'Permanently delete Active Recording?',
  label: 'Delete Active Recording',
  description: `Recording and report data will be lost.`,
  ariaLabel: 'Recording delete warning',
};

export const DeleteArchivedRecordings: DeleteWarning = {
  id: DeleteWarningType.DeleteArchivedRecordings,
  title: 'Permanently delete Archived Recording?',
  label: 'Delete Archived Recording',
  description: `Recording and report data will be lost.`,
  ariaLabel: 'Recording delete warning',
};

export const DeleteAutomatedRules: DeleteWarning = {
  id: DeleteWarningType.DeleteAutomatedRules,
  title: 'Permanently delete Automated Rule?',
  label: 'Delete Automated Rule',
  description: `Rule data will be lost.`,
  ariaLabel: 'Automated rule delete warning',
};

export const DeleteEventTemplates: DeleteWarning = {
  id: DeleteWarningType.DeleteEventTemplates,
  title: 'Permanently delete Event Template?',
  label: 'Delete Event Template',
  description: `Custom event template data will be lost.`,
  ariaLabel: 'Event template delete warning',
};

export const DeleteProbeTemplates: DeleteWarning = {
  id: DeleteWarningType.DeleteProbeTemplates,
  title: 'Permanently delete Probe Template?',
  label: 'Delete Probe Template',
  description: `Custom probe template data will be lost.`,
  ariaLabel: 'Probe template delete warning',
};

export const DeleteActiveProbes: DeleteWarning = {
  id: DeleteWarningType.DeleteActiveProbes,
  title: 'Permanently remove Active Probes from the target?',
  label: 'Remove Active Probes',
  description: `Active probes will be removed from the target.`,
  ariaLabel: 'Active Probes remove warning',
};

export const DeleteJMXCredentials: DeleteWarning = {
  id: DeleteWarningType.DeleteJMXCredentials,
  title: 'Permanently delete JMX Credentials?',
  label: 'Delete JMX Credentials',
  description: `Credential data for this target will be lost.`,
  ariaLabel: 'JMX Credentials delete warning',
};

export const DeleteCustomTargets: DeleteWarning = {
  id: DeleteWarningType.DeleteCustomTargets,
  title: 'Permanently delete Custom Target?',
  label: 'Delete Custom Targets',
  description: `Custom target information will be lost.`,
  ariaLabel: 'Custom Targets delete warning',
};

export const DeleteWarningKinds: DeleteWarning[] = [
  DeleteActiveRecordings,
  DeleteArchivedRecordings,
  DeleteAutomatedRules,
  DeleteEventTemplates,
  DeleteProbeTemplates,
  DeleteActiveProbes,
  DeleteJMXCredentials,
  DeleteCustomTargets,
];

export const getFromWarningMap = (warning: DeleteWarningType): DeleteWarning | undefined => {
  const wt = DeleteWarningKinds.find((t) => t.id === warning);
  return wt;
};
