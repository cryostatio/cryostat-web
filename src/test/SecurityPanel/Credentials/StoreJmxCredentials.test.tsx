/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { cleanup, screen, within } from '@testing-library/react';
import { of, throwError } from 'rxjs';
import { MatchedCredential, StoredCredential } from '@app/Shared/Services/Api.service';
import { Modal, ModalVariant } from '@patternfly/react-core';
import { NotificationMessage } from '@app/Shared/Services/NotificationChannel.service';
import { StoreJmxCredentials } from '@app/SecurityPanel/Credentials/StoreJmxCredentials';
import { ServiceContext, defaultServices } from '@app/Shared/Services/Services';
import { DeleteJMXCredentials, DeleteWarningType } from '@app/Modal/DeleteWarningUtils';
import { Target } from '@app/Shared/Services/Target.service';
import { TargetDiscoveryEvent } from '@app/Shared/Services/Targets.service';

import { renderWithServiceContext } from '../../Common';
import { CreateJmxCredentialModalProps } from '@app/SecurityPanel/Credentials/CreateJmxCredentialModal';
import { NotificationsContext, NotificationsInstance } from '@app/Notifications/Notifications';

const mockCredential: StoredCredential = {
  id: 0,
  matchExpression: 'target.connectUrl == "service:jmx:rmi://someUrl"',
  numMatchingTargets: 1,
};
const mockAnotherCredential: StoredCredential = {
  id: 1,
  matchExpression:
    'target.connectUrl == "service:jmx:rmi://anotherUrl" || target.connectUrl == "service:jmx:rmi://anotherMatchUrl" || target.connectUrl == "service:jmx:rmi://yetAnotherMatchUrl"',
  numMatchingTargets: 2,
};

const mockTarget: Target = { connectUrl: 'service:jmx:rmi://someUrl', alias: 'someAlias' };
const mockAnotherTarget: Target = { connectUrl: 'service:jmx:rmi://anotherUrl', alias: 'anotherAlias' };
const mockAnotherMatchingTarget: Target = {
  connectUrl: 'service:jmx:rmi://anotherMatchUrl',
  alias: 'anotherMatchAlias',
};
const mockYetAnotherMatchingTarget: Target = {
  connectUrl: 'service:jmx:rmi://yetAnotherMatchUrl',
  alias: 'yetAnotherMatchAlias',
};

const mockMatchedCredentialResponse: MatchedCredential = {
  matchExpression: mockCredential.matchExpression,
  targets: [mockTarget],
};
const mockAnotherMatchedCredentialResponse: MatchedCredential = {
  matchExpression: mockAnotherCredential.matchExpression,
  targets: [mockAnotherTarget, mockAnotherMatchingTarget],
};

const mockCredentialNotification = { message: mockCredential } as NotificationMessage;
const evt = { kind: 'LOST', serviceRef: mockTarget } as TargetDiscoveryEvent;

const mockLostTargetNotification = {
  message: { event: { kind: 'LOST', serviceRef: mockAnotherTarget } },
} as NotificationMessage;

const mockFoundTargetNotification = {
  message: { event: { kind: 'FOUND', serviceRef: mockYetAnotherMatchingTarget } },
} as NotificationMessage;

jest.mock('@app/SecurityPanel/Credentials/CreateJmxCredentialModal', () => {
  return {
    CreateJmxCredentialModal: jest.fn((props: CreateJmxCredentialModalProps) => {
      return (
        <Modal
          isOpen={props.visible}
          variant={ModalVariant.large}
          showClose={true}
          onClose={props.onDismiss}
          title="CreateJmxCredentialModal"
        >
          Jmx Auth Form
        </Modal>
      );
    }),
  };
});

jest
  .spyOn(defaultServices.notificationChannel, 'messages')
  .mockReturnValueOnce(of()) // 'renders correctly'
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of(mockCredentialNotification)) // 'adds the correct table entry when a stored notification is received'
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // 'removes the correct table entry when a deletion notification is received'
  .mockReturnValueOnce(of(mockCredentialNotification))
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // 'renders an empty table after receiving deletion notifications for all credentials'
  .mockReturnValueOnce(of(mockCredentialNotification))
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // 'expands to show the correct nested targets'
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of())

  .mockReturnValueOnce(of()) // 'decrements the correct count and updates the correct nested table when a lost target notification is received'
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockLostTargetNotification))
  .mockReturnValueOnce(of(mockLostTargetNotification))
  .mockReturnValueOnce(of(mockLostTargetNotification))

  .mockReturnValueOnce(of()) // 'increments the correct count and updates the correct nested table when a found target notification is received'
  .mockReturnValueOnce(of())
  .mockReturnValueOnce(of(mockFoundTargetNotification))
  .mockReturnValueOnce(of(mockFoundTargetNotification))
  .mockReturnValueOnce(of(mockFoundTargetNotification))

  .mockReturnValue(of()), // remaining tests
  jest.spyOn(defaultServices.api, 'deleteCredentials').mockReturnValue(of(true));
jest
  .spyOn(defaultServices.api, 'getCredentials')
  .mockReturnValueOnce(of([mockCredential, mockAnotherCredential])) // 'renders correctly'

  .mockReturnValueOnce(of([])) // 'adds the correct table entry when a stored notification is received'

  .mockReturnValueOnce(of([mockCredential, mockAnotherCredential])) // 'removes the correct table entry when a deletion notification is received'

  .mockReturnValueOnce(of([mockCredential])) // 'renders an empty table after receiving deletion notifications for all credentials'

  .mockReturnValueOnce(of([mockCredential, mockAnotherCredential])) // 'expands to show the correct nested targets'

  .mockReturnValueOnce(of([mockCredential, mockAnotherCredential])) // 'decrements the correct count and updates the correct nested table when a lost target notification is received'

  .mockReturnValueOnce(of([mockCredential, mockAnotherCredential])) // 'increments the correct count and updates the correct nested table when a found target notification is received'

  .mockReturnValueOnce(of([])) // 'opens the JMX auth modal when Add is clicked'

  .mockReturnValueOnce(of([mockCredential, mockAnotherCredential])) // 'shows a popup when Delete is clicked and makes a delete request when deleting one credential after confirming Delete'

  .mockReturnValueOnce(of([mockCredential, mockAnotherCredential])) // 'makes multiple delete requests when all credentials are deleted at once'

  .mockReturnValue(throwError(() => new Error('Too many calls')));

jest
  .spyOn(defaultServices.api, 'getCredential')
  .mockReturnValueOnce(of(mockMatchedCredentialResponse)) // 'expands to show the correct nested targets'
  .mockReturnValueOnce(of(mockAnotherMatchedCredentialResponse))

  .mockReturnValueOnce(of(mockMatchedCredentialResponse)) // 'decrements the correct count and updates the correct nested table when a lost target notification is received'
  .mockReturnValueOnce(of(mockAnotherMatchedCredentialResponse))

  .mockReturnValueOnce(of(mockMatchedCredentialResponse)) // 'increments the correct count and updates the correct nested table when a found target notification is received'
  .mockReturnValueOnce(of(mockAnotherMatchedCredentialResponse))

  .mockReturnValue(throwError(() => new Error('Too many calls')));

jest.spyOn(defaultServices.settings, 'deletionDialogsEnabledFor').mockReturnValueOnce(true);

describe('<StoreJmxCredentials />', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    const apiRequestSpy = jest.spyOn(defaultServices.api, 'getCredentials');
    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <StoreJmxCredentials />
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();

    expect(apiRequestSpy).toHaveBeenCalledTimes(1);
  });

  it('adds the correct table entry when a stored notification is received', async () => {
    const apiRequestSpy = jest.spyOn(defaultServices.api, 'getCredentials');
    renderWithServiceContext(<StoreJmxCredentials />);

    expect(screen.getByText(mockCredential.matchExpression)).toBeInTheDocument();
    expect(screen.getByText(mockCredential.numMatchingTargets)).toBeInTheDocument();
    expect(apiRequestSpy).toHaveBeenCalledTimes(1);
  });

  it('removes the correct table entry when a deletion notification is received', async () => {
    const apiRequestSpy = jest.spyOn(defaultServices.api, 'getCredentials');
    renderWithServiceContext(<StoreJmxCredentials />);

    expect(screen.queryByText(mockCredential.matchExpression)).not.toBeInTheDocument();
    expect(screen.queryByText(mockCredential.numMatchingTargets)).not.toBeInTheDocument();
    expect(screen.getByText(mockAnotherCredential.matchExpression)).toBeInTheDocument();
    expect(screen.getByText(mockAnotherCredential.numMatchingTargets)).toBeInTheDocument();
    expect(apiRequestSpy).toHaveBeenCalledTimes(1);
  });

  it('renders an empty table after receiving deletion notifications for all credentials', async () => {
    const apiRequestSpy = jest.spyOn(defaultServices.api, 'getCredentials');
    renderWithServiceContext(<StoreJmxCredentials />);

    expect(screen.queryByText(mockCredential.matchExpression)).not.toBeInTheDocument();
    expect(screen.queryByText(mockCredential.numMatchingTargets)).not.toBeInTheDocument();
    expect(screen.queryByText(mockAnotherCredential.matchExpression)).not.toBeInTheDocument();
    expect(screen.queryByText(mockAnotherCredential.numMatchingTargets)).not.toBeInTheDocument();
    expect(screen.getByText('No Stored Credentials')).toBeInTheDocument();
    expect(apiRequestSpy).toHaveBeenCalledTimes(1);
  });

  it('expands to show the correct nested targets', async () => {
    const { user } = renderWithServiceContext(<StoreJmxCredentials />);

    expect(screen.queryByText('Target')).not.toBeInTheDocument();
    expect(screen.queryByText(`${mockTarget.alias} (${mockTarget.connectUrl})`)).not.toBeInTheDocument();
    expect(screen.queryByText(`${mockAnotherTarget.alias} (${mockAnotherTarget.connectUrl})`)).not.toBeInTheDocument();
    expect(
      screen.queryByText(`${mockAnotherMatchingTarget.alias} (${mockAnotherMatchingTarget.connectUrl})`)
    ).not.toBeInTheDocument();

    const expandButtons = screen.getAllByLabelText('Details');

    await user.click(expandButtons[0]);

    expect(screen.getByText('Target')).toBeInTheDocument();
    expect(screen.getByText(`${mockTarget.alias} (${mockTarget.connectUrl})`)).toBeInTheDocument();
    expect(screen.queryByText(`${mockAnotherTarget.alias} (${mockAnotherTarget.connectUrl})`)).not.toBeInTheDocument();
    expect(
      screen.queryByText(`${mockAnotherMatchingTarget.alias} (${mockAnotherMatchingTarget.connectUrl})`)
    ).not.toBeInTheDocument();

    await user.click(expandButtons[1]);

    expect(screen.getByText(`${mockAnotherTarget.alias} (${mockAnotherTarget.connectUrl})`)).toBeInTheDocument();
    expect(
      screen.getByText(`${mockAnotherMatchingTarget.alias} (${mockAnotherMatchingTarget.connectUrl})`)
    ).toBeInTheDocument();
  });

  it('decrements the correct count and updates the correct nested table when a lost target notification is received', async () => {
    const { user } = renderWithServiceContext(<StoreJmxCredentials />);

    // both counts should now be equal to 1
    const counts = screen.getAllByText(mockAnotherCredential.numMatchingTargets - 1);
    expect(within(counts[0]).getByText(mockCredential.numMatchingTargets)).toBeTruthy();
    expect(within(counts[1]).getByText(mockAnotherCredential.numMatchingTargets - 1)).toBeTruthy();

    const expandButtons = screen.getAllByLabelText('Details');

    await user.click(expandButtons[0]);

    expect(screen.getByText(`${mockTarget.alias} (${mockTarget.connectUrl})`)).toBeInTheDocument();

    await user.click(expandButtons[1]);

    expect(screen.queryByText(`${mockAnotherTarget.alias} (${mockAnotherTarget.connectUrl})`)).not.toBeInTheDocument();
    expect(
      screen.getByText(`${mockAnotherMatchingTarget.alias} (${mockAnotherMatchingTarget.connectUrl})`)
    ).toBeInTheDocument();
  });

  it('increments the correct count and updates the correct nested table when a found target notification is received', async () => {
    const { user } = renderWithServiceContext(<StoreJmxCredentials />);

    expect(screen.getByText(mockCredential.numMatchingTargets)).toBeInTheDocument();
    expect(screen.getByText(mockAnotherCredential.numMatchingTargets + 1)).toBeInTheDocument();

    const expandButtons = screen.getAllByLabelText('Details');

    await user.click(expandButtons[0]);

    expect(screen.getByText(`${mockTarget.alias} (${mockTarget.connectUrl})`)).toBeInTheDocument();
    expect(
      screen.queryByText(`${mockYetAnotherMatchingTarget.alias} (${mockYetAnotherMatchingTarget.connectUrl})`)
    ).not.toBeInTheDocument();

    await user.click(expandButtons[1]);

    expect(screen.getByText(`${mockAnotherTarget.alias} (${mockAnotherTarget.connectUrl})`)).toBeInTheDocument();
    expect(
      screen.getByText(`${mockAnotherMatchingTarget.alias} (${mockAnotherMatchingTarget.connectUrl})`)
    ).toBeInTheDocument();
    expect(
      screen.getByText(`${mockYetAnotherMatchingTarget.alias} (${mockYetAnotherMatchingTarget.connectUrl})`)
    ).toBeInTheDocument();
  });

  it('opens the JMX auth modal when Add is clicked', async () => {
    const { user } = renderWithServiceContext(<StoreJmxCredentials />);

    await user.click(screen.getByText('Add'));
    expect(screen.getByText('CreateJmxCredentialModal')).toBeInTheDocument();

    await user.click(screen.getByLabelText('Close'));
    expect(screen.queryByText('CreateJmxCredentialModal')).not.toBeInTheDocument();
  });

  it('shows a popup when Delete is clicked and makes a delete request when deleting one credential after confirming Delete', async () => {
    const queryRequestSpy = jest.spyOn(defaultServices.api, 'getCredentials');
    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'deleteCredentials');
    const { user } = renderWithServiceContext(<StoreJmxCredentials />);

    expect(screen.getByText(mockCredential.matchExpression)).toBeInTheDocument();
    expect(screen.getByText(mockAnotherCredential.matchExpression)).toBeInTheDocument();

    await user.click(screen.getByLabelText('credentials-table-row-0-check'));
    await user.click(screen.getByText('Delete'));

    expect(screen.getByLabelText(DeleteJMXCredentials.ariaLabel)).toBeInTheDocument();

    const dialogWarningSpy = jest.spyOn(defaultServices.settings, 'setDeletionDialogsEnabledFor');
    await user.click(screen.getByLabelText("Don't ask me again"));
    await user.click(within(screen.getByLabelText(DeleteJMXCredentials.ariaLabel)).getByText('Delete'));

    expect(dialogWarningSpy).toBeCalledTimes(1);
    expect(dialogWarningSpy).toBeCalledWith(DeleteWarningType.DeleteJMXCredentials, false);
    expect(queryRequestSpy).toHaveBeenCalledTimes(1);
    expect(deleteRequestSpy).toHaveBeenCalledTimes(1);
  });

  it('makes multiple delete requests when all credentials are deleted at once w/o popup warning', async () => {
    const queryRequestSpy = jest.spyOn(defaultServices.api, 'getCredentials');
    const deleteRequestSpy = jest.spyOn(defaultServices.api, 'deleteCredentials');
    const { user } = renderWithServiceContext(<StoreJmxCredentials />);

    expect(screen.getByText(mockCredential.matchExpression)).toBeInTheDocument();
    expect(screen.getByText(mockCredential.numMatchingTargets)).toBeInTheDocument();
    expect(screen.getByText(mockAnotherCredential.matchExpression)).toBeInTheDocument();
    expect(screen.getByText(mockAnotherCredential.numMatchingTargets)).toBeInTheDocument();

    const checkboxes = screen.getAllByRole('checkbox');
    const selectAllCheck = checkboxes[0];
    await user.click(selectAllCheck);
    await user.click(screen.getByText('Delete'));
    expect(queryRequestSpy).toHaveBeenCalledTimes(1);
    expect(deleteRequestSpy).toHaveBeenCalledTimes(2);
  });
});
