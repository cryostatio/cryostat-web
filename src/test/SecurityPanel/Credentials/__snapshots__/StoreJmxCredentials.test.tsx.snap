// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StoreJmxCredentials /> renders correctly 1`] = `
Array [
  <div
    className="pf-c-toolbar"
    data-ouia-component-id="OUIA-Generated-Toolbar-2"
    data-ouia-component-type="PF4/Toolbar"
    data-ouia-safe={true}
    id="target-credentials-toolbar"
  >
    <div
      className="pf-c-toolbar__content"
    >
      <div
        className="pf-c-toolbar__content-section"
      >
        <div
          className="pf-c-toolbar__item"
        >
          <button
            aria-disabled={false}
            aria-label="add-jmx-credential"
            className="pf-c-button pf-m-primary"
            data-ouia-component-id="OUIA-Generated-Button-primary-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={false}
            onClick={[Function]}
            role={null}
            type="button"
          >
            Add
          </button>
        </div>
        <div
          className="pf-c-toolbar__item"
        >
          <button
            aria-disabled={true}
            aria-label="delete-selected-jmx-credential"
            className="pf-c-button pf-m-danger pf-m-disabled"
            data-ouia-component-id="OUIA-Generated-Button-danger-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={true}
            onClick={[Function]}
            role={null}
            tabIndex={null}
            type="button"
          >
            Delete
          </button>
        </div>
      </div>
      <div
        className="pf-c-toolbar__expandable-content"
        id="target-credentials-toolbar-expandable-content-1"
      >
        <div
          className="pf-c-toolbar__group"
        />
      </div>
    </div>
    <div
      className="pf-c-toolbar__content pf-m-hidden"
      hidden={true}
    >
      <div
        className="pf-c-toolbar__group"
      />
    </div>
  </div>,
  <table
    aria-label="Stored Credentials"
    className="pf-c-table pf-m-grid-md"
    data-ouia-component-id="OUIA-Generated-Table-1"
    data-ouia-component-type="PF4/Table"
    data-ouia-safe={true}
    role="grid"
  >
    <thead
      className=""
    >
      <tr
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-1"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <th
          className=""
          onMouseEnter={[Function]}
          scope={null}
        />
        <td
          className="pf-c-table__check"
          onMouseEnter={[Function]}
          scope={null}
        >
          <label>
            <input
              aria-label="Select all rows"
              checked={false}
              name="check-all"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </td>
        <th
          className="pf-m-width-90"
          onMouseEnter={[Function]}
          scope="col"
        >
          Match Expression
        </th>
        <th
          className="pf-m-width-15"
          onMouseEnter={[Function]}
          scope="col"
        >
          Count
        </th>
      </tr>
    </thead>
    <tbody
      className=""
      role="rowgroup"
    >
      <tr
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-2"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <td
          aria-controls="credentials-ex-expand-0"
          className="pf-c-table__toggle"
          id="credentials-ex-toggle-0"
          onMouseEnter={[Function]}
        >
          <button
            aria-disabled={false}
            aria-expanded={false}
            aria-label="Details"
            aria-labelledby="simple-node0 expand-toggle0"
            className="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={false}
            id="expand-toggle0"
            onClick={[Function]}
            role={null}
            type="button"
          >
            <div
              className="pf-c-table__toggle-icon"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  Object {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                />
              </svg>
            </div>
          </button>
        </td>
        <td
          className=""
          onMouseEnter={[Function]}
        >
          <div
            className="pf-c-check pf-m-standalone"
          >
            <input
              aria-invalid={false}
              aria-label="credentials-table-row-0-check"
              checked={false}
              className="pf-c-check__input"
              data-ouia-component-id="OUIA-Generated-Checkbox-1"
              data-ouia-component-type="PF4/Checkbox"
              data-ouia-safe={true}
              disabled={false}
              id="credentials-table-row-0-check"
              name="credentials-table-row-0-check"
              onChange={[Function]}
              required={false}
              type="checkbox"
            />
          </div>
        </td>
        <td
          className=""
          data-label="Match Expression"
          onMouseEnter={[Function]}
        >
          target.connectUrl == "service:jmx:rmi://someUrl"
        </td>
        <td
          className=""
          data-label="Count"
          onMouseEnter={[Function]}
        >
          <span
            className="pf-c-badge pf-m-unread"
          >
            1
          </span>
        </td>
      </tr>
      <tr
        className="pf-c-table__expandable-row"
        data-ouia-component-id="OUIA-Generated-TableRow-3"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={true}
      >
        <td
          className=""
          colSpan={4}
          data-label="Content Details"
          onMouseEnter={[Function]}
        />
      </tr>
      <tr
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-4"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <td
          aria-controls="credentials-ex-expand-1"
          className="pf-c-table__toggle"
          id="credentials-ex-toggle-1"
          onMouseEnter={[Function]}
        >
          <button
            aria-disabled={false}
            aria-expanded={false}
            aria-label="Details"
            aria-labelledby="simple-node1 expand-toggle1"
            className="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={false}
            id="expand-toggle1"
            onClick={[Function]}
            role={null}
            type="button"
          >
            <div
              className="pf-c-table__toggle-icon"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  Object {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                />
              </svg>
            </div>
          </button>
        </td>
        <td
          className=""
          onMouseEnter={[Function]}
        >
          <div
            className="pf-c-check pf-m-standalone"
          >
            <input
              aria-invalid={false}
              aria-label="credentials-table-row-1-check"
              checked={false}
              className="pf-c-check__input"
              data-ouia-component-id="OUIA-Generated-Checkbox-2"
              data-ouia-component-type="PF4/Checkbox"
              data-ouia-safe={true}
              disabled={false}
              id="credentials-table-row-1-check"
              name="credentials-table-row-1-check"
              onChange={[Function]}
              required={false}
              type="checkbox"
            />
          </div>
        </td>
        <td
          className=""
          data-label="Match Expression"
          onMouseEnter={[Function]}
        >
          target.connectUrl == "service:jmx:rmi://anotherUrl" || target.connectUrl == "service:jmx:rmi://anotherMatchUrl" || target.connectUrl == "service:jmx:rmi://yetAnotherMatchUrl"
        </td>
        <td
          className=""
          data-label="Count"
          onMouseEnter={[Function]}
        >
          <span
            className="pf-c-badge pf-m-unread"
          >
            2
          </span>
        </td>
      </tr>
      <tr
        className="pf-c-table__expandable-row"
        data-ouia-component-id="OUIA-Generated-TableRow-5"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={true}
      >
        <td
          className=""
          colSpan={4}
          data-label="Content Details"
          onMouseEnter={[Function]}
        />
      </tr>
    </tbody>
  </table>,
]
`;
